File: src/test/java/com/esentri/wrabbitmq/java/ExceptionTest.java
package com.esentri.wrabbitmq.java;

import com.esentri.wrabbitmq.WrabbitEventWithReply;
import com.esentri.wrabbitmq.WrabbitTopic;
import com.esentri.wrabbitmq.exceptions.WrabbitReplyBasicException;
import com.esentri.wrabbitmq.exceptions.WrabbitReplyTimeoutException;
import com.esentri.wrabbitmq.exceptions.WrabbitSerializationException;
import com.esentri.wrabbitmq.java.testhelper.TestException;
import com.esentri.wrabbitmq.testhelper.NonSerializableException;
import org.junit.jupiter.api.Test;

import java.io.Serializable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;

import static com.esentri.wrabbitmq.java.TestFunctions.Await;
import static com.esentri.wrabbitmq.java.TestFunctions.NewEventWithReply;
import static org.fest.assertions.Assertions.assertThat;

public class ExceptionTest {

   private WrabbitTopic topic = new WrabbitTopic("TestTopic-Exception");

   private <MESSAGE extends Serializable, REPLY extends Serializable> WrabbitEventWithReply<MESSAGE, REPLY> newEventWithReply() {
      return NewEventWithReply(topic);
   }

   @Test
   public void noReplier_withDefaultTimeout() throws ExecutionException, InterruptedException {
      WrabbitEventWithReply<String, String> event = newEventWithReply();
      event.sendAndReceive("hello").handle((v, e) -> {
              assertThat(e.getCause()).isInstanceOf(WrabbitReplyTimeoutException.class);
              assertThat(e.getMessage())
                 .contains(event.getEventName())
                 .contains(topic.getTopicName());
              return null;
           }).join();
   }

   @Test
   public void noReplier_withCustomTimeout() {
      WrabbitEventWithReply<String, String> event = newEventWithReply();
      event.sendAndReceive("hello", 10).handle((v, e) -> {
              assertThat(e.getCause()).isInstanceOf(WrabbitReplyTimeoutException.class);
              assertThat(e.getMessage())
                 .contains(event.getEventName())
                 .contains(topic.getTopicName());
              return null;
           }).join();
   }

   @Test
   public void replierThrowsException() {
      WrabbitEventWithReply<String, String> event = newEventWithReply();
      String exceptionMessage = "hello world";

      event.replier(it -> {
         throw new TestException(exceptionMessage);
      });

      event.sendAndReceive("hello").handle((v, e) -> {
              assertThat(e.getCause()).isInstanceOf(WrabbitReplyBasicException.class);
              assertThat(e.getCause().getCause()).isInstanceOf(TestException.class);
              assertThat(e.getCause().getCause()).hasMessage(exceptionMessage);
              return null;
           }).join();
   }

   @Test
   public void nonSerializableExceptionTest() {
      WrabbitEventWithReply<String, String> event = newEventWithReply();
      event.replier (it -> {
         throw new NonSerializableException();
      });
      event.sendAndReceive("hello", 10000).handle((v, e) -> {
         assertThat(e.getCause()).isInstanceOf(WrabbitReplyBasicException.class);
         assertThat(e.getCause().getCause()).isInstanceOf(WrabbitSerializationException.class);
         assertThat(e.getCause().getCause().getCause()).isNull();
         return null;
      }).join();
   }

}

File: src/test/kotlin/com/esentri/wrabbitmq/ExceptionTest.kt
package com.esentri.wrabbitmq

import com.esentri.wrabbitmq.exceptions.WrabbitReplyBasicException
import com.esentri.wrabbitmq.exceptions.WrabbitReplyTimeoutException
import com.esentri.wrabbitmq.testhelper.CustomException
import com.esentri.wrabbitmq.testhelper.TestException
import org.fest.assertions.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.io.Serializable

class ExceptionTest {

   private val topic = WrabbitTopic("TestTopic-Exception")
   private fun <MESSAGE : Serializable, REPLY : Serializable> newEventWithReply() =
      NewEventWithReply<MESSAGE, REPLY>(topic)

   @Test
   fun noReplier_withDefaultTimeout() {
      val event = newEventWithReply<String, String>()
      event.sendAndReceive("hello").handle {_, e ->
         assertThat(e.cause).isInstanceOf(WrabbitReplyTimeoutException::class.java)
         assertThat(e.message)
            .contains(event.eventName)
            .contains(topic.topicName)
      }.get()
   }

   @Test
   fun noReplier_withCustomTimeout() {
      val event = newEventWithReply<String, String>()
      event.sendAndReceive("hello", 10).handle {_, e ->
         assertThat(e.cause).isInstanceOf(WrabbitReplyTimeoutException::class.java)
         assertThat(e.message)
            .contains(event.eventName)
            .contains(topic.topicName)
      }.get()
   }

   @Test
   fun replierThrowsException() {
      val exceptionMessage = "hello world"
      val event = newEventWithReply<String, String>()
      event.replier { _ ->
         throw TestException(exceptionMessage)
      }
      event.sendAndReceive("hello", 10000).handle {_, e ->
         assertThat(e.cause).isInstanceOf(WrabbitReplyBasicException::class.java)
         assertThat(e.cause!!.cause).isInstanceOf(TestException::class.java)
         assertThat(e.cause!!.cause).hasMessage(exceptionMessage)
      }.get()
   }

   @Test
   fun customException() {
      val exceptionMessage = "custom exception"
      val event = newEventWithReply<String, String>()
      event.replier { _ ->
         throw CustomException(exceptionMessage)
      }
      event.sendAndReceive("hello", 10000).handle {_, e ->
         assertThat(e.cause).isInstanceOf(WrabbitReplyBasicException::class.java)
         assertThat(e.cause!!.cause).isInstanceOf(CustomException::class.java)
         assertThat(e.cause!!.cause).hasMessage(exceptionMessage)
      }.get()
   }

}